allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,707 @@
+    feature_size: ``int``
+        The embedding size for all the embedded features, such as distances or span widths.
+    max_span_width: ``int``
+        The maximum width of candidate spans.
+    lexical_dropout: ``int``
+        The probability of dropping out dimensions of the embedded text.
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 context_layer: Seq2SeqEncoder,
+                 mention_feedforward: FeedForward,
+                 antecedent_feedforward: FeedForward,
+                 feature_size: int,
+                 max_span_width: int,
+                 spans_per_word: float,
+                 max_antecedents: int,
+                 lexical_dropout: float = 0.2,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+        super(CoreferenceResolver, self).__init__(vocab, regularizer)
+
+        self._text_field_embedder = text_field_embedder
+        self._context_layer = context_layer
+        self._mention_feedforward = TimeDistributed(mention_feedforward)
+        self._antecedent_feedforward = TimeDistributed(antecedent_feedforward)
+        self._mention_scorer = TimeDistributed(torch.nn.Linear(mention_feedforward.get_output_dim(), 1))
+        self._antecedent_scorer = TimeDistributed(torch.nn.Linear(antecedent_feedforward.get_output_dim(), 1))
+        self._head_scorer = TimeDistributed(torch.nn.Linear(context_layer.get_output_dim(), 1))
+
+        # 10 possible distance buckets.
+        self._distance_embedding = Embedding(10, feature_size)
+        self._span_width_embedding = Embedding(max_span_width, feature_size)
+
+        self._max_span_width = max_span_width
+        self._spans_per_word = spans_per_word
+        self._max_antecedents = max_antecedents
+
+        self._mention_recall = MentionRecall()
+        self._conll_coref_scores = ConllCorefScores()
+        if lexical_dropout > 0:
+            self._lexical_dropout = torch.nn.Dropout(p=lexical_dropout)
+        else:
+            self._lexical_dropout = lambda x: x
+        initializer(self)
+
+    def _compute_head_attention(self,
+                                head_scores: torch.FloatTensor,
+                                text_embeddings: torch.FloatTensor,
+                                span_ends: torch.IntTensor,
+                                span_size: torch.IntTensor):
+        """
+        Parameters
+        ----------
+        head_scores : torch.FloatTensor
+            Unnormalized attention scores for every word. This score is shared for every
```

Comment:
Just out of curiosity, how did you expect the attention weights to be parametrized?

Reply:
I would have expected some kind of span-dependent scores for heads (yours _is_, but only through the normalization).  After thinking more about the options here, I think your modeling decision is a good one, for both statistical and computational efficiency, it just surprised me to see it written this way when I first read it, particularly because there was no text describing the decision.